sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -409,6 +505,340 @@ class GpuDynamicPartitionDataWriter(
   }
 }
 
+/**
+ * Dynamic partition writer with concurrent writers, meaning multiple concurrent writers are opened
+ * for writing.
+ *
+ * The process has the following steps:
+ *  - Step 1: Maintain a map of output writers per each partition columns. Keep all
+ *    writers opened; Cache the inputted batches by splitting them into sub-groups and
+ *    each partition holds a list of sub-groups; Find and write the the max pending
+ *    partition data if the total caches exceed the limitation.
+ *  - Step 2: If number of concurrent writers exceeds limit, sort rest of rows on partition.
+ *    Write rows one by one, and eagerly close the writer when finishing
+ *  Caller is expected to call `writeWithIterator()` instead of `write()` to write records.
+ */
+class GpuDynamicPartitionDataConcurrentWriter(
+    description: GpuWriteJobDescription,
+    taskAttemptContext: TaskAttemptContext,
+    committer: FileCommitProtocol,
+    spec: GpuConcurrentOutputWriterSpec)
+    extends GpuDynamicPartitionDataSingleWriter(description, taskAttemptContext, committer)
+        with Arm with Logging {
+
+  // Keep all the unclosed writers, key is partition directory string.
+  // Note: if fall back to sort-based mode, also use the opened writers in the map.
+  private val concurrentWriters = mutable.HashMap[String, WriterStatusWithCaches]()
+
+  // guarantee to close the caches and writers when task is finished
+  TaskContext.get().addTaskCompletionListener[Unit](_ => closeCachesAndWriters())
+
+  private val outDataTypes = description.dataColumns.map(_.dataType).toArray
+
+  override def abort(): Unit = {
+    try {
+      closeCachesAndWriters()
+    } finally {
+      committer.abortTask(taskAttemptContext)
+    }
+  }
+
+  /** Wrapper class for status of a unique concurrent output writer. */
+  private class WriterStatusWithCaches(
+      // writer status
+      var writerStatus: WriterStatus,
+
+      // caches for this partition or writer
+      val tableCaches: ListBuffer[Table] = ListBuffer(),
+
+      // current device bytes for the above caches
+      var deviceBytes: Long = 0
+  )
+
+  /**
+   * State to indicate if we are falling back to sort-based writer.
+   * Because we first try to use concurrent writers, its initial value is false.
+   */
+  private var fallBackToSortBased: Boolean = false
+
+  def writeWithSingleWriter(cb: ColumnarBatch): Unit = {
+    // invoke `GpuDynamicPartitionDataSingleWriter`.write
+    super.write(cb, partDir => concurrentWriters.get(partDir).map(e => e.writerStatus))
```

Comment:
Is it ever possible for us to have cached data and this ends up skipping that data? Could we at least add some comments about how this should not be called unless all of the data in the cache is already written out?

Reply:
Added comments.
Now the `single writer` should handle the un-closed writers and should handle the un-flushed caches.
When falling back to `single writer`,  `single writer` will concatenate the un-flushed caches and original sorted partition data to avoid tiny groups.